Display all languages in the contest in teaminterface

vmcj opened the pull request:
Also limit the languages in case the problem has different languages.

<img width="2787" height="1152" alt="image" src="https://github.com/user-attachments/assets/2ddca29d-dfa7-4d2c-aab4-a709b1c30fc3" />

I don't really like the `[All]` as it is not really true but `rest` is also not that good. Found this when playing around with something and needed the info so added it.

@nickygerritsen suggested this as alternative:
<img width="2181" height="1665" alt="image" src="https://github.com/user-attachments/assets/28aad487-9875-4f0c-91e2-87e6ce83c270" />
Or this:
<img width="2263" height="834" alt="image" src="https://github.com/user-attachments/assets/9f7b0080-afde-4944-a105-dd594b4d6dfc" />

vmcj committed:
Display all problems on the language overview for teams

Also limit the languages in case the problem has different languages.

vmcj commented:
> I like the idea, but I would not do `All`, but:
> 
>     * Not display any badge anywhere if we have no problem specific languages. (you already do this I think)
Yes

> 
>     * Display exact problewms otherwise. So Instead of `All` display a badge for each problem that does not exclude it.
I didn't try that but thought it would be overcrowded, but I can try it.

meisterT commented:
I like adding this info in general, wonder if we need something special for too many problems (e.g. > 15) like online judges where this different languages per problem are more likely

meisterT commented on webapp/src/Controller/Team/LanguageController.php:
```diff
@@ -33,16 +34,73 @@ public function __construct(
 
+    /**
+     * @param Language[] $languages
+     * @return Language[]
```
This is not the return type that we actually return, is it?

vmcj replied:
Yes, no idea why I didn't document that properly, fixed now.

vmcj committed:
For team submit show all possible languages in the contest

We could have specific languages for a single problem which are not allowed in the whole contest.

In the submit page we now list all specific languages for any problem in the contest and the
allowed contest languages if there is no specific limitation for any of the problems.

vmcj committed:
Pass for which problem a language is allowed for the overview page

This way we now have on the team/problemset page which languages are allowed,
and on the language page for which problems it's allowed.

vmcj committed:
Repair the page

vmcj committed:
Fix word wrapping for extension

In case you have an unreasonable amount of extensions the wrapping would
not happen and the list would flow out of the card. Alternative is to use the
`overflow: auto` to add a scrollbar

vmcj commented:
Extreme case
<img width="1470" height="784" alt="image" src="https://github.com/user-attachments/assets/84f0ba8c-5dc4-45e7-96ee-19f3bb4b7ac3" />

Normal case:
<img width="1470" height="518" alt="image" src="https://github.com/user-attachments/assets/58f4243d-b9b6-439c-a9ba-0bfba0904666" />

I think I like the 2nd placement the most, so left aligned and no scroll but it would make it more crowded for the online judge scenario

meisterT commented:
Can you please do the screenshot of the full page (or at least full box for one language)?

vmcj committed:
Use my preferance

vmcj committed:
Show same overview on Problemset page

This shows all allowed languages instead of only the restricted to a problem.

Before we could have a problem which was restricted and get the Language field but we would
display nothing for the other problems

vmcj commented:
<img width="1956" height="902" alt="image" src="https://github.com/user-attachments/assets/5d12f541-ac80-4c1e-8324-548f33a9c759" />

Not 100% sure if this is an improvement but it would provide all needed information instead of forcing the team to think about this themselves.

vmcj commented on webapp/src/Controller/Team/LanguageController.php:
```diff
@@ -33,16 +34,73 @@ public function __construct(
+     */
+    private function addLanguage(array $languages, Language $language, ContestProblem $problem, bool $inverted = false): array
+    {
+        $langId = $language->getName();
```
No,

vmcj committed:
Revert "Use my preferance"

This reverts commit eec1e1808e9fd6e9a16096f5556974cb6ca8fe34.

vmcj commented:
> Can you please do the screenshot of the full page (or at least full box for one language)?

<img width="2536" height="1578" alt="image" src="https://github.com/user-attachments/assets/864cd789-eaed-444d-9f0d-c66bf24db47c" />
<img width="2536" height="1578" alt="image" src="https://github.com/user-attachments/assets/ad8e4e9f-6c30-424d-914c-fcd4d2e3b73b" />

meisterT commented on webapp/src/Utils/Utils.php:
```diff
@@ -1147,4 +1148,13 @@ public static function numericToBcMath(float|int|string $value, int $scale = 9):
+     * @param Set<V> $b
+     */
+    public static function equalSets(Set $a, Set $b): bool {
+        return $a->diff($b)->isEmpty() && $b->diff($a)->isEmpty();
```
consider using this instead:
```
return $a->xor($b)->isEmpty();
```

vmcj replied:
Would we still need the helper method or can it be taken out?

meisterT commented on webapp/src/Controller/Team/LanguageController.php:
```diff
@@ -40,7 +58,32 @@ public function languagesAction(): Response
+            $allLanguages->add($language);
+        }
+        // Add the problem specific languages
+        foreach ($this->dj->getCurrentContest()->getProblems() as $problem) {
```
`getCurrentContest` can return null, can't it? Also consider using the local var `$currentContest` that you create above

meisterT reviewed (commented):
we have something like "collect all languages including problem-specific ones" in a couple places, consider putting some helper method in DOMjudgeService

vmcj committed:
Handle the case where there is no active contest

vmcj committed:
Use more elegant compare for sets

vmcj committed:
Fix PHPStan issue

vmcj committed:
Extract shared code

vmcj committed:
Do some extraction

vmcj commented:
> we have something like "collect all languages including problem-specific ones" in a couple places, consider putting some helper method in DOMjudgeService

I tried, but from the 3 places 2 need extra information to return the right format. So I'm not sure if it will make the code better.

meisterT commented on webapp/src/Service/DOMJudgeService.php:
```diff
@@ -1177,6 +1178,20 @@ public function getTwigDataForProblemsAction(
+            foreach ($problems as $problem) {
+                if (!count($problem->getProblem()->getLanguages()) && $limited) {
+                    foreach ($this->getAllowedLanguagesForContest($contest) as $language) {
+                        $problem->getProblem()->addLanguage($language);
```
I am not particularly fond of mutating the problem's internal state (calling `addLanguage`) in a method which is called `getSomething`.

meisterT commented on webapp/src/Service/DOMJudgeService.php:
```diff
@@ -1177,6 +1178,20 @@ public function getTwigDataForProblemsAction(
+            $limited = false;
+            foreach ($problems as $problem) {
+                if (count($problem->getProblem()->getLanguages())) {
+                    $limited = true;
```
Is this `$limited` dead code?

meisterT commented on webapp/src/Service/DOMJudgeService.php:
```diff
@@ -1847,9 +1862,25 @@ public function shadowMode(): bool
+                    }
+                }
+                if ($limited) {
+                    $allLanguages = $allLanguages->merge($languages);
```
Where is `$allLanguages` used?

meisterT replied:
Looks like the whole `if ($alsoProblemSpecific)` branch is dead code?

meisterT commented on webapp/src/Controller/Team/LanguageController.php:
```diff
@@ -32,6 +36,20 @@ public function __construct(
+        if (!isset($languages[$langId])) {
+            $languages[$langId] = ['problems' => [], 'language' => $language];
+        }
+        $languages[$langId]['problems'][] = $problem;
```
Is that what we want if `$problem` is `null`?

meisterT commented on webapp/templates/team/languages.html.twig:
```diff
@@ -7,19 +7,46 @@
+                                    {% if limited %}
+                                        <div style="margin-left: auto; float: right; text-align: right;">
+                                            {% for problem in problems %}
+                                                {{ problem | problemBadge }}
```
Not sure which of these three you want to keep. Leaving the comment just to make sure we remember to pick one option.
